File: Leetcode/1. Two Sum.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */

// Array && HashTable
// Map object holds key-value pair
// Map.has();
// Map.get();
// Map.set();

var twoSum = (nums, target) => {
    const dict = new Map();
    for (let i = 0; i < nums.length; i++) {
        let complement = target- nums[i];
        if (dict.has(complement)) {
            return [dict.get(complement), i];
        }
        dict.set(nums[i], i);
    }
};

// Traverse over the array containing n element only once, Time Complexity is O(n);
// Space Complexity is O(n);

File: Leetcode-Java/1. Two Sum.java
class Solution {
    public int[] twoSum(int[] nums, int target) {
        Map<Integer, Integer> map = new HashMap<>();
        for (int i = 0; i < nums.length; i++) {
            int complement = target - nums[i];
            if (map.containsKey(complement)) {
                return new int[] {
                  map.get(complement), i
                };
            }
            map.put(nums[i], i);
        }
        throw new IllegalArgumentException("No Two Sum Solution");
    }
}

Time Complexity: O(n); // We traverse the list containing n elements only once. Each look up in the table costs only O(1) time.
Space Complexity: O(n); // The extra space required depends on the number of items stored in hash table, which stores at most n elements.
